File: wpimath/src/main/native/cpp/geometry/proto/Rectangle2dProto.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/geometry/proto/Rectangle2dProto.hpp"

#include "wpi/util/protobuf/ProtobufCallbacks.hpp"
#include "wpimath/protobuf/geometry2d.npb.h"

std::optional<wpi::math::Rectangle2d> wpi::util::Protobuf<wpi::math::Rectangle2d>::Unpack(
    InputStream& stream) {
  wpi::util::UnpackCallback<wpi::math::Pose2d> pose;
  wpi_proto_ProtobufRectangle2d msg{
      .center = pose.Callback(),
      .xWidth = 0,
      .yWidth = 0,
  };
  if (!stream.Decode(msg)) {
    return {};
  }

  auto ipose = pose.Items();

  if (ipose.empty()) {
    return {};
  }

  return wpi::math::Rectangle2d{
      ipose[0],
      wpi::units::meter_t{msg.xWidth},
      wpi::units::meter_t{msg.yWidth},
  };
}

bool wpi::util::Protobuf<wpi::math::Rectangle2d>::Pack(OutputStream& stream,
                                           const wpi::math::Rectangle2d& value) {
  wpi::util::PackCallback pose{&value.Center()};
  wpi_proto_ProtobufRectangle2d msg{
      .center = pose.Callback(),
      .xWidth = value.XWidth().value(),
      .yWidth = value.YWidth().value(),
  };
  return stream.Encode(msg);
}

File: wpimath/src/main/java/org/wpilib/math/geometry/proto/Rectangle2dProto.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.geometry.proto;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rectangle2d;
import edu.wpi.first.math.proto.Geometry2D.ProtobufRectangle2d;
import edu.wpi.first.util.protobuf.Protobuf;
import us.hebi.quickbuf.Descriptors.Descriptor;

public class Rectangle2dProto implements Protobuf<Rectangle2d, ProtobufRectangle2d> {
  @Override
  public Class<Rectangle2d> getTypeClass() {
    return Rectangle2d.class;
  }

  @Override
  public Descriptor getDescriptor() {
    return ProtobufRectangle2d.getDescriptor();
  }

  @Override
  public ProtobufRectangle2d createMessage() {
    return ProtobufRectangle2d.newInstance();
  }

  @Override
  public Rectangle2d unpack(ProtobufRectangle2d msg) {
    return new Rectangle2d(Pose2d.proto.unpack(msg.getCenter()), msg.getXWidth(), msg.getYWidth());
  }

  @Override
  public void pack(ProtobufRectangle2d msg, Rectangle2d value) {
    Pose2d.proto.pack(msg.getMutableCenter(), value.getCenter());
    msg.setXWidth(value.getXWidth());
    msg.setYWidth(value.getYWidth());
  }
}
